### PR TITLE
Refactor plot and plot3d to use virtualfile_from_data

### LINKED ISSUE #1021: Support 2D list as input
**Description of the problem**

PyGMT doesn't allow 2D list as input.

**Full code that generated the error**

The script below doesn't work:
```python
import pygmt
fig = pygmt.Figure()

fig.plot(data=[[0, 0]], region=[-2, 2, -2, 2], frame=True, style="c0.5c", pen="1p")
fig.show()
```

but it works if I convert it to a 2D numpy array:
```python
import pygmt
fig = pygmt.Figure()

fig.plot(data=np.array([[0, 0]]), region=[-2, 2, -2, 2], frame=True, style="c0.5c", pen="1p")
fig.show()
```

**Full error message**

```
---------------------------------------------------------------------------
GMTInvalidInput                           Traceback (most recent call last)
<ipython-input-15-cf008b39b139> in <module>
      2 fig = pygmt.Figure()
      3 
----> 4 fig.plot([[0, 0]], region=[-2, 2, -2, 2], frame=True, style="c0.5c", pen="1p")
      5 fig.show()

~/Gits/gmt/pygmt/pygmt/helpers/decorators.py in new_module(*args, **kwargs)
    286                 if alias in kwargs:
    287                     kwargs[arg] = kwargs.pop(alias)
--> 288             return module_func(*args, **kwargs)
    289 
    290         new_module.aliases = aliases

~/Gits/gmt/pygmt/pygmt/helpers/decorators.py in new_module(*args, **kwargs)
    430                         kwargs[arg] = separators[fmt].join(f"{item}" for item in value)
    431             # Execute the original function and return its output
--> 432             return module_func(*args, **kwargs)
    433 
    434         return new_module

~/Gits/gmt/pygmt/pygmt/src/plot.py in plot(self, x, y, data, sizes, direction, **kwargs)
    202     kwargs = self._preprocess(**kwargs)  # pylint: disable=protected-access
    203 
--> 204     kind = data_kind(data, x, y)
    205 
    206     extra_arrays = []

~/Gits/gmt/pygmt/pygmt/helpers/utils.py in data_kind(data, x, y, z)
     62         raise GMTInvalidInput("Too much data. Use either data or x and y.")
     63     if data is None and (x is None or y is None):
---> 64         raise GMTInvalidInput("Must provided both x and y.")
     65 
     66     if isinstance(data, str):

GMTInvalidInput: Must provided both x and y.
```

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Added an additional `extra_arrays` parameter to the`virtualfile_from_data` function to accept optional numpy arrays from the plot and plot3d functions.

~~**TODO**: May need to tweak the `data_kind` function a little, to output `matrix` for true numpy array matrices and `vectors` for numpy arrays with mixed dtypes. This is to fix a segmentation fault on `test_plot3d_matrix_color`.~~ Edit: see comment at https://github.com/GenericMappingTools/pygmt/pull/990#discussion_r586959742

Part of #949. Once this is done, it will simplify the  `geopandas` integration (#608) (which will be done in a follow-up Pull Request at #1000).

Also fixes #1021 

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/clib/session.py [hunk 1]
```diff
@@ -1430,14 +1435,25 @@ def virtualfile_from_data(self, check_kind=None, data=None, x=None, y=None, z=No
...
-                _data = np.atleast_2d(np.asanyarray(data).T)
+                try:
+                    # Just use virtualfile_from_matrix for 2D
+                    # numpy.ndarray which are not datetime (M) types
+                    assert data.ndim == 2 and not data.dtype.kind == "M"
+                    _virtualfile_from = self.virtualfile_from_matrix
+                    _data = (data,)
+                except (AssertionError, AttributeError):
+                    # Python lists, tuples, and numpy ndarray types
+                    _data = np.atleast_2d(np.asanyarray(data).T)
```
Comment (weiji14) [inline pygmt/clib/session.py:1456]: Would appreciate advice on improving/refactoring this chunk of code. I'm half thinking whether to move some of the logic to `pygmt/helpers/utils.py`, i.e. have a new 'kind' besides file/grid/matrix/vectors:

https://github.com/GenericMappingTools/pygmt/blob/65b099a5873b4f9e76d7fccdbc8882d9b37870e3/pygmt/helpers/utils.py#L65-L72

Or we could just keep things like this as it is.

Comment (seisman) [reply pygmt/clib/session.py:1456]: I didn't read the codes carefully, but I think the "matrix" kind is quite confusing. 

The GMT API function `GMT_Put_Matrix()`, the PyGMT wrapper `put_matrix()` and the virtualfile function `virtualfile_from_matrix()` all require a simple 2d matrix with a single dtype (e.g., `np.float` or `np.double`). 

However, currently, data types like pandas.DataFrame are also "matrix". So we have to check the data types to choose either `virtualfile_from_vector` or `virtualfile_from_matrix`.

I agree with you that we can/shoudl add a new kind to distinguish between a GMT-compatible "matrix" and a more complicated data structure.

Comment (weiji14) [reply pygmt/clib/session.py:1456]: How about we keep this logic as is for now to resolve #1021 (which helps with #1020), and refactor things later to have more specific 'matrix' types?

Comment (seisman) [reply pygmt/clib/session.py:1456]: Sounds good.

File: pygmt/clib/session.py [hunk 2]
```diff
@@ -1378,6 +1380,9 @@ def virtualfile_from_data(self, check_kind=None, data=None, x=None, y=None, z=No
             raster grid, a vector matrix/arrays, or other supported data input.
         x/y/z : 1d arrays or None
             x, y and z columns as numpy arrays.
+        extra_arrays : list
```
Comment (seisman) [inline pygmt/clib/session.py:1383]:
```suggestion
        extra_arrays : list of 1d arrays
```

File: pygmt/clib/session.py [hunk 3]
```diff
@@ -1430,14 +1435,25 @@ def virtualfile_from_data(self, check_kind=None, data=None, x=None, y=None, z=No
...
             try:
                 # pandas.DataFrame and xarray.Dataset types
                 _data = [array for _, array in data.items()]
             except AttributeError:
-                # Python lists, tuples, and numpy ndarray types
-                _data = np.atleast_2d(np.asanyarray(data).T)
+                try:
+                    # Just use virtualfile_from_matrix for 2D
+                    # numpy.ndarray which are not datetime (M) types
+                    assert data.ndim == 2 and not data.dtype.kind == "M"
```
Comment (seisman) [inline pygmt/clib/session.py:1451]: > assert data.ndim == 2 and not data.dtype.kind == "M"

The [`GMT_Put_Matrix()`](https://docs.generic-mapping-tools.org/dev/api.html#gmt-put-matrix) function only supports a few [numeric data types](https://docs.generic-mapping-tools.org/dev/api.html#tbl-types), but [numpy.dtype.kind](https://numpy.org/doc/stable/reference/generated/numpy.dtype.kind.html) can have values other than `M`, for example, `data.dtype.kind == "O"` will pass the assert statement but is not supported by `virtualfile_from_matrix`. Perhaps check `data.dtype.kind in 'iuf'`?

**Comment by seisman** [comment]
One more comment: perhaps we should add a new test to make sure #1021 is fixed.

File: pygmt/clib/session.py (hunk 3 above)
Comment (weiji14) [reply pygmt/clib/session.py:1451]: Ok, I'll try `data.dtype.kind in 'iuf'` and see if it works on the test suite.

**Review by seisman — APPROVED** [review]
OK to merge after all tests pass and the styling issue is fixed.

**Pull request merged**